### ISSUE TITLE
GCC warns -Wnon-template-friend on LLVM

### ISSUE THREAD

**Issue by danakj** [llvm/llvm-project#67942]
Clang does [not have this warning](https://github.com/llvm/llvm-project/issues/23716) but GCC is finding this mistake in LLVM 18 due to a recent change.

```
/usr/lib/llvm-18/include/llvm/ADT/PagedVector.h:213:59: warning: friend declaration ‘bool llvm::operator==(const PagedVector<T, PageSize>::MaterializedIterator&, const PagedVector<T, PageSize>::MaterializedIterator&)’ declares a non-template function [-Wnon-template-friend]
  213 |                            MaterializedIterator const &RHS);
      |                                                           ^
/usr/lib/llvm-18/include/llvm/ADT/PagedVector.h:213:59: note: (if this is not what you intended, make sure the function template has already been declared and add ‘<>’ after the function name here)
/usr/lib/llvm-18/include/llvm/ADT/PagedVector.h:215:59: warning: friend declaration ‘bool llvm::operator!=(const PagedVector<T, PageSize>::MaterializedIterator&, const PagedVector<T, PageSize>::MaterializedIterator&)’ declares a non-template function [-Wnon-template-friend]
  215 |                            MaterializedIterator const &RHS);
      |                                                           ^
```

**Comment by shafik** [comment]
CC @ktf @zygoloid @AaronBallman

**Comment by ktf** [comment]
I am not sure how to fix that, frankly. If I drop the friend forward declaration in MaterializedIterator, the error goes away on Linux, but then XCode 15 complains with:

```
In file included from /Users/ktf/src/sw/SOURCES/llvm-upstream/master/0/llvm/unittests/ADT/PagedVectorTest.cpp:13:
/Users/ktf/src/sw/SOURCES/llvm-upstream/master/0/llvm/include/llvm/ADT/PagedVector.h:224:16: error: 'PV' is a private member of 'llvm::PagedVector<int, 10>::MaterializedIterator'
    assert(LHS.PV == RHS.PV);
               ^
/Users/ktf/src/sw/SOURCES/llvm-upstream/master/0/llvm/include/llvm/ADT/PagedVector.h:240:18: note: in instantiation of member function 'llvm::operator==' requested here
    return !(LHS == RHS);
                 ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/c++/v1/__iterator/distance.h:35:20: note: in instantiation of member function
'llvm::operator!=' requested here
    for (; __first != __last; ++__first)
                   ^
/Users/ktf/src/sw/SOURCES/llvm-upstream/master/0/llvm/include/llvm/ADT/PagedVector.h:175:24: note: implicitly declared private here
    const PagedVector *PV;
                       ^
/Users/ktf/src/sw/SOURCES/llvm-upstream/master/0/llvm/include/llvm/ADT/PagedVector.h:224:26: error: 'PV' is a private member of 'llvm::PagedVector<int, 10>::MaterializedIterator'
    assert(LHS.PV == RHS.PV);
                         ^
/Users/ktf/src/sw/SOURCES/llvm-upstream/master/0/llvm/include/llvm/ADT/PagedVector.h:175:24: note: implicitly declared private here
    const PagedVector *PV;
                       ^
```

if I inline the operators like in https://github.com/llvm/llvm-project/pull/67958, it works on macOS and Linux (both clang and GCC), however I think I saw it failing in the Windows CI.
